templates/17_upsert.tpl

```diff
@@ -73,36 +72,64 @@ func (o *{{$tableNameSingular}}) Upsert(exec boil.Executor, {{if ne .DriverName
 	var err error
 
 	if !cached {
-		var ret []string
-		whitelist, ret = strmangle.InsertColumnSet(
+		insert, ret := strmangle.InsertColumnSet(
 			{{$varNameSingular}}Columns,
 			{{$varNameSingular}}ColumnsWithDefault,
 			{{$varNameSingular}}ColumnsWithoutDefault,
 			nzDefaults,
 			whitelist,
 		)
+		{{- if eq .DriverName "mssql"}}
+		insert = strmangle.SetComplement(insert, {{$varNameSingular}}ColumnsWithAuto)
+		for i, v := range insert {
+			if strmangle.ContainsAny({{$varNameSingular}}PrimaryKeyColumns, v) && strmangle.ContainsAny({{$varNameSingular}}ColumnsWithDefault, v) {
+				insert = append(insert[:i], insert[i+1:]...)
+			}
+		}
+		if len(insert) == 0 {
+			return errors.New("models: unable to upsert {{.Table.Name}}, could not build insert column list")
+		}
+
+		ret = strmangle.SetMerge(ret, {{$varNameSingular}}ColumnsWithAuto)
+		ret = strmangle.SetMerge(ret, {{$varNameSingular}}ColumnsWithDefault)
+		
+		{{end -}}
 		update := strmangle.UpdateColumnSet(
 			{{$varNameSingular}}Columns,
 			{{$varNameSingular}}PrimaryKeyColumns,
 			updateColumns,
 		)
+
+		{{- if eq .DriverName "mssql"}}
+		update = strmangle.SetComplement(update, {{$varNameSingular}}ColumnsWithAuto)
+		{{end -}}
+
 		if len(update) == 0 {
 			return errors.New("{{.PkgName}}: unable to upsert {{.Table.Name}}, could not build update column list")
 		}
 
-		{{if ne .DriverName "mysql" -}}
+		{{if eq .DriverName "postgres"}}
 		conflict := conflictColumns
 		if len(conflict) == 0 {
 			conflict = make([]string, len({{$varNameSingular}}PrimaryKeyColumns))
 			copy(conflict, {{$varNameSingular}}PrimaryKeyColumns)
 		}
-		cache.query = queries.BuildUpsertQueryPostgres(dialect, "{{$schemaTable}}", updateOnConflict, ret, update, conflict, whitelist)
-		{{- else -}}
-		cache.query = queries.BuildUpsertQueryMySQL(dialect, "{{.Table.Name}}", update, whitelist)
+		cache.query = queries.BuildUpsertQueryPostgres(dialect, "{{$schemaTable}}", updateOnConflict, ret, update, conflict, insert)
+		{{- end -}}
+
+		{{if eq .DriverName "mysql"}}
```

Comment:
Possible to clean this section up with `{{else if ...}}` instead of several if/ends.